src/swap/useSwapQuote.ts

```diff
@@ -1,27 +1,238 @@
   estimatedPriceImpact: BigNumber | null
+  preparedTransactions?: PreparedTransactionsResult
+}
+
+interface PreparedTransactionsPossible {
+  type: 'possible'
+  approveTransaction: TransactionRequestCIP42 & { amountToApprove: bigint }
+  swapTransaction: TransactionRequestCIP42
+}
+
+export interface PreparedTransactionsNeedDecreaseSwapAmountForGas {
+  type: 'need-decrease-swap-amount-for-gas'
+  maxGasCost: BigNumber
+  feeCurrency: TokenBalance
+  decreasedSwapAmount: BigNumber
+}
+
+export interface PreparedTransactionsNotEnoughBalanceForGas {
+  type: 'not-enough-balance-for-gas'
+  feeCurrencies: TokenBalance[]
+}
+
+export type PreparedTransactionsResult =
+  | PreparedTransactionsPossible
+  | PreparedTransactionsNeedDecreaseSwapAmountForGas
+  | PreparedTransactionsNotEnoughBalanceForGas
+
+function createBaseApproveTransaction(
+  updatedField: Field,
+  unvalidatedSwapTransaction: SwapTransaction
+) {
+  const { guaranteedPrice, sellTokenAddress, buyAmount, sellAmount, allowanceTarget, from } =
+    unvalidatedSwapTransaction
+  const amountType: string =
+    updatedField === Field.TO ? ('buyAmount' as const) : ('sellAmount' as const)
+
+  const amountToApprove =
+    amountType === 'buyAmount'
+      ? BigInt(new BigNumber(buyAmount).times(guaranteedPrice).toFixed(0, 0))
+      : BigInt(sellAmount)
+
+  const data = encodeFunctionData({
+    abi: erc20.abi,
+    functionName: 'approve',
+    args: [allowanceTarget as Address, amountToApprove],
+  })
+
+  const approveTx: TransactionRequestCIP42 & { amountToApprove: bigint } = {
+    from: from as Address,
+    to: sellTokenAddress as Address,
+    data,
+    type: 'cip42',
+    amountToApprove,
+  }
+
+  return approveTx
+}
+
+function createBaseSwapTransaction(unvalidatedSwapTransaction: SwapTransaction) {
+  const { from, to, value, data, gas } = unvalidatedSwapTransaction
+
+  const swapTx: TransactionRequestCIP42 & { gas: bigint } = {
+    from: from as Address,
+    to: to as Address,
+    value: BigInt(value ? value : 0),
+    data: data as Hex,
+    gas: BigInt(gas), // This isn't entirely accurate
+    type: 'cip42',
+  }
+
+  return swapTx
+}
+
+function getMaxGasCost(txs: TransactionRequestCIP42[]) {
+  let maxGasCost = BigInt(0)
+  for (const tx of txs) {
+    if (!tx.gas || !tx.maxFeePerGas) {
+      throw new Error('Missing gas or maxFeePerGas')
+    }
+    maxGasCost += BigInt(tx.gas) * BigInt(tx.maxFeePerGas)
+  }
+  return maxGasCost
+}
+
+async function prepareTransactions(
+  fromToken: TokenBalanceWithAddress,
+  updatedField: Field,
+  unvalidatedSwapTransaction: SwapTransaction,
+  price: string,
+  feeCurrencies: TokenBalance[]
+) {
+  const baseApproveTx = createBaseApproveTransaction(updatedField, unvalidatedSwapTransaction)
+  const baseSwapTx = createBaseSwapTransaction(unvalidatedSwapTransaction)
+
+  const maxGasCosts: Array<{ feeCurrency: TokenBalance; maxGasCostInDecimal: BigNumber }> = []
+  for (const feeCurrency of feeCurrencies) {
```

Comment:
this function is pretty long and complex. could the stuff in the for loop be pulled out into a helper function to make it easier to read, and to write tests for?

Reply:
Definitely agree. This code is at the "make it work" stage and I haven't tried to break it down differently yet. I know we probably want to make it more generic, but until we have other examples to better guide this, I preferred to wait and see.